tilavarauspalvelu/integrations/keyless_entry/client.py

```diff
@@ -41,124 +51,374 @@ class PindoraClient(BaseExternalServiceClient):
+            response,
+            reservation,
+            action="updating reservation",
+            expected_status_code=HTTP_204_NO_CONTENT,
+        )
 
     @classmethod
     def delete_reservation(cls, reservation: Reservation) -> None:
-        """Delete reservation from Pindora."""
+        """Delete a reservation from Pindora."""
         url = cls._build_url(f"reservation/{reservation.ext_uuid}")
 
         response = cls.delete(url=url)
-        cls._validate_reservation_response(response, reservation)
-
-        if response.status_code != HTTP_204_NO_CONTENT:
-            msg = (
-                f"Unexpected response from Pindora when deleting reservation '{reservation.ext_uuid}': "
-                f"[{response.status_code}] {response.text}"
-            )
-            raise PindoraAPIError(msg)
+        cls._validate_reservation_response(
+            response,
+            reservation,
+            action="deleting reservation",
+            expected_status_code=HTTP_204_NO_CONTENT,
+        )
 
     @classmethod
     def change_reservation_access_code(cls, reservation: Reservation) -> PindoraReservationResponse:
-        """Change reservation access code in Pindora."""
+        """Change a reservation's access code in Pindora."""
         url = cls._build_url(f"change-access-code/reservation/{reservation.ext_uuid}")
 
         response = cls.put(url=url)
-        cls._validate_response(response)
+        cls._validate_reservation_response(
+            response,
+            reservation,
+            action="changing access code for reservation",
+        )
 
-        if response.status_code == HTTP_404_NOT_FOUND:
-            msg = f"Reservation '{reservation.ext_uuid}' not found from Pindora."
+        data = cls.response_json(response)
+        return cls._parse_reservation_response(data)
+
+    ####################
+    # Seasonal booking #
+    ####################
+
+    @classmethod
+    def get_seasonal_booking(cls, application_section: ApplicationSection) -> PindoraSeasonalBookingResponse:
+        """Fetch a seasonal booking from Pindora."""
+        url = cls._build_url(f"seasonal-booking/{application_section.ext_uuid}")
+
+        response = cls.get(url=url)
+        cls._validate_seasonal_booking_response(
+            response,
+            application_section,
+            action="fetching seasonal booking",
+        )
+
+        data = cls.response_json(response)
+        return cls._parse_seasonal_booking_response(data)
+
+    @classmethod
+    def create_seasonal_booking(
+        cls,
+        application_section: ApplicationSection,
+        *,
+        is_active: bool = False,
+    ) -> PindoraSeasonalBookingResponse:
+        """Create a new seasonal booking in Pindora."""
+        url = cls._build_url("seasonal-bookings")
+
+        reservations: list[Reservation] = list(
+            application_section.actions.get_reservations().select_related("recurring_reservation__reservation_unit")
```

Comment:
Only `CONFIRMED` reservations

Reply:
Just confirmed or also those that are [going to occur](https://github.com/City-of-Helsinki/tilavarauspalvelu-core/blob/8f18ba31f98074afbbd426c1f266e1a5264b8031/tilavarauspalvelu/enums.py#L436-L442)?

Reply:
I think it will have to be just CONFIRMED since individual reservations in the other states might not, in the end, actually occur. The reservee (applicant) could in such a case then end up with an access code that also works for that particular occurrence, if the other three states were included and assumed to take place.
In practical terms, these states are not really supposed to exist in the series/seasonal context and the the individual reservations are CONFIRMED to begin with (unless some seasonal ones are DENIED before the application round results are sent to the applicants).